youtube_dl/extractor/bilibili.py

```diff
@@ -120,141 +120,346 @@ def _report_error(self, result):
         else:
             raise ExtractorError('Can\'t extract Bangumi episode ID')
 
+    def _getfps(self,s:str)->int:
+        "convert fps to int"
+        if s.isnumeric() :
+            return int(s)
+        else :
+            r=re.search(r"([0-9]+)/([0-9]+)",s)
+            if r!=None :
+                r=r.groups()
+                return int(r[0])/int(r[1])
+            else :
+                return 0
+    
+    def _calculate_size(self,durl:list) -> int :
+        "Calculate total file size."
+        s=0
+        for i in durl :
+            s=s+i['size']
+        return s
+
     def _real_extract(self, url):
         url, smuggled_data = unsmuggle_url(url, {})
 
         mobj = re.match(self._VALID_URL, url)
         video_id = mobj.group('id') or mobj.group('id_bv')
-        anime_id = mobj.group('anime_id')
+        query=compat_urlparse.parse_qs(compat_urlparse.urlparse(url).query)
+        part=None
+        if 'p' in query and str(query['p'][0]).isnumeric() :
+            part=int(query['p'][0])
+
+        #Set Cookies need to parse the Links.
+        self._set_cookie(domain=".bilibili.com",name="CURRENT_QUALITY",value="120")#Set default video quality
+        self._set_cookie(domain=".bilibili.com",name="CURRENT_FNVAL",value="16")
+        self._set_cookie(domain=".bilibili.com",name="laboratory",value="1-1")#Use new webpage API
+        self._set_cookie(domain=".bilibili.com",name="stardustvideo",value="1")
+
         webpage = self._download_webpage(url, video_id)
 
-        if 'anime/' not in url:
-            cid = self._search_regex(
-                r'\bcid(?:["\']:|=)(\d+)', webpage, 'cid',
-                default=None
-            ) or compat_parse_qs(self._search_regex(
-                [r'EmbedPlayer\([^)]+,\s*"([^"]+)"\)',
-                 r'EmbedPlayer\([^)]+,\s*\\"([^"]+)\\"\)',
-                 r'<iframe[^>]+src="https://secure\.bilibili\.com/secure,([^"]+)"'],
-                webpage, 'player parameters'))['cid'][0]
-        else:
-            if 'no_bangumi_tip' not in smuggled_data:
-                self.to_screen('Downloading episode %s. To download all videos in anime %s, re-run youtube-dl with %s' % (
-                    video_id, anime_id, compat_urlparse.urljoin(url, '//bangumi.bilibili.com/anime/%s' % anime_id)))
-            headers = {
-                'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8',
-                'Referer': url
-            }
-            headers.update(self.geo_verification_headers())
-
-            js = self._download_json(
-                'http://bangumi.bilibili.com/web_api/get_source', video_id,
-                data=urlencode_postdata({'episode_id': video_id}),
-                headers=headers)
-            if 'result' not in js:
-                self._report_error(js)
-            cid = js['result']['cid']
-
-        headers = {
-            'Referer': url
+        video_info=re.search(r"window\.__INITIAL_STATE__=([^;]+)",webpage,re.I)
```

Comment:
The regexp can be problematic when the json payload contains ";", for example the anime title can be "Dat Girl;）". In this case the matched result is not a valid json string.